### PR TITLE
Added an SPF Check button to the config tab for the client

### PR DESCRIPTION
**Pull request by guitarmanj**
@wolfthefallen @zeroSteiner

### CODE REVIEW

**Comment by zeroSteiner** [comment]
The changes to the requirements file can definitely not be merged. They'd remove a couple of necessary packages.

File: king_phisher/client/tabs/mail.py [hunk 1]
```diff
@@ -847,6 +847,37 @@ def signal_button_clicked_verify(self, button):
...
+			return True
+
+		if not spf_result:
+			gui_utilities.show_dialog_info('No SPF records found.\n', self.parent)
+		else:
+			if spf_result is 'fail':
+				gui_utilities.show_dialog_info('SPF exists with a hard fail.\nYour messages will probably be blocked.\n', self.parent)
+			elif spf_result is 'softfail':
+				gui_utilities.show_dialog_info('SPF Exists with a soft fail.\nYour messages have strong possiblity of being blocked. Check your logs.\n', self.parent)
+			return True
```
Comment (zeroSteiner) [inline king_phisher/client/tabs/mail.py]: This line is unnecessary.

File: king_phisher/client/tabs/mail.py [hunk 2]
```diff
@@ -847,6 +847,37 @@ def signal_button_clicked_verify(self, button):
 			gui_utilities.show_dialog_info('Successfully Opened The Web Server URL', self.parent)
 		return
 
+	def signal_button_clicked_verify_spf(self, button):
+		sender_email = self.gobjects['entry_source_email_smtp'].get_text()
+		error_description = None
+
+		if not utilities.is_valid_email_address(sender_email):
+			gui_utilities.show_dialog_info('WARNING: Can not check SPF records for an invalid source email address.\n', self.parent)
```
Comment (zeroSteiner) [inline king_phisher/client/tabs/mail.py]: This should use show_dialog_warning instead so the WARNING prefix isn't necessary.

File: king_phisher/client/tabs/mail.py [hunk 3]
```diff
@@ -847,6 +847,37 @@ def signal_button_clicked_verify(self, button):
 			gui_utilities.show_dialog_info('Successfully Opened The Web Server URL', self.parent)
 		return
 
+	def signal_button_clicked_verify_spf(self, button):
+		sender_email = self.gobjects['entry_source_email_smtp'].get_text()
+		error_description = None
```
Comment (zeroSteiner) [inline king_phisher/client/tabs/mail.py]: I don't see where this is referenced.

File: king_phisher/client/tabs/mail.py [hunk 4]
```diff
@@ -847,6 +847,37 @@ def signal_button_clicked_verify(self, button):
...
+
+		if not utilities.is_valid_email_address(sender_email):
+			gui_utilities.show_dialog_info('WARNING: Can not check SPF records for an invalid source email address.\n', self.parent)
+			return True
+
+		spf_test_ip = mailer.guess_smtp_server_address(self.config['smtp_server'], (self.config['ssh_server'] if self.config['smtp_ssh_enable'] else None))
+		if not spf_test_ip:
+			gui_utilities.show_dialog_info('skipping spf policy check because the smtp server address could not be reliably detected', self.parent)
```
Comment (zeroSteiner) [inline king_phisher/client/tabs/mail.py]: This should also probably be a warning since the user presumably cares about checking SPF.

**Comment by zeroSteiner** [comment]
The dialogs would all look a bit better if a title were specified. A title can be provided as a third string argument. That way all the result messages could have a common title, same with the warnings etc.

File: king_phisher/client/tabs/mail.py (hunk 3 above)
Comment (guitarmanj) [inline king_phisher/client/tabs/mail.py]: The sender_email variable is referenced on lines 854 and 863.

Comment (zeroSteiner) [inline king_phisher/client/tabs/mail.py]: I meant the `error_description` variable, is that one used? I had originally reviewed this on my cell so it wasn't the easiest way to review things.

Comment (guitarmanj) [inline king_phisher/client/tabs/mail.py]: fixed

File: requirements.txt [hunk 5]
```diff
@@ -12,11 +12,11 @@ icalendar>=3.9.2
 ipaddress>=1.0.16              ; python_version < '3.3'
 Jinja2>=2.8
 markupsafe>=0.23
-matplotlib>=1.5.1
+#matplotlib>=1.5.1
```
Comment (wolfthefallen) [inline requirements.txt]: matplotlib is a requirement for King Phisher, it should not be commented out.

File: requirements.txt [hunk 6]
```diff
@@ -12,11 +12,11 @@ icalendar>=3.9.2
...
 msgpack-python>=0.4.7
 paramiko>=2.0.0
 pluginbase>=0.4
-psycopg2>=2.6.1
+#psycopg2>=2.6.1
```
Comment (wolfthefallen) [inline requirements.txt]: same, this PR does not require any changes to the lib requirements of King Phisher.

File: king_phisher/client/tabs/mail.py [hunk 7]
```diff
@@ -847,6 +847,36 @@ def signal_button_clicked_verify(self, button):
...
+		spf_test_sender, spf_test_domain = sender_email.split('@')
+		try:
+			spf_test = spf.SenderPolicyFramework(spf_test_ip, spf_test_domain, spf_test_sender)
+			spf_result = spf_test.check_host()
+		except spf.SPFError as error:
+			gui_utilities.show_dialog_warning('Warning', self.parent,"Done, encountered exception: {0}.\n".format(error.__class__.__name__))
+			return True
+
+		if not spf_result:
+			gui_utilities.show_dialog_info('SPF Check Results', self.parent, 'No SPF records found.\n')
```
Comment (wolfthefallen) [inline king_phisher/client/tabs/mail.py]: \n not required, no second line of text.

File: king_phisher/client/tabs/mail.py [hunk 8]
```diff
@@ -847,6 +847,36 @@ def signal_button_clicked_verify(self, button):
...
+		if not spf_test_ip:
+			gui_utilities.show_dialog_warning('Warning', self.parent, 'Skipping spf policy check because the smtp server address could not be reliably detected')
+			return True
+
+		spf_test_sender, spf_test_domain = sender_email.split('@')
+		try:
+			spf_test = spf.SenderPolicyFramework(spf_test_ip, spf_test_domain, spf_test_sender)
+			spf_result = spf_test.check_host()
+		except spf.SPFError as error:
+			gui_utilities.show_dialog_warning('Warning', self.parent,"Done, encountered exception: {0}.\n".format(error.__class__.__name__))
```
Comment (wolfthefallen) [inline king_phisher/client/tabs/mail.py]: \n not required, no second line of text

File: king_phisher/client/tabs/mail.py [hunk 9]
```diff
@@ -847,6 +847,36 @@ def signal_button_clicked_verify(self, button):
 			gui_utilities.show_dialog_info('Successfully Opened The Web Server URL', self.parent)
 		return
 
+	def signal_button_clicked_verify_spf(self, button):
+		sender_email = self.gobjects['entry_source_email_smtp'].get_text()
+		
+		if not utilities.is_valid_email_address(sender_email):
+			gui_utilities.show_dialog_warning('Warning', self.parent, 'Can not check SPF records for an invalid source email address.\n')
```
Comment (wolfthefallen) [inline king_phisher/client/tabs/mail.py]: \n not required, no second line of text

**Comment by wolfthefallen** [comment]
@guitarmanj looks good, Thank you. Once I test, if everything works out i'll land it.

Need to test
- [ ] spf check fail
 - [ ] spf check soft fail
 - [ ] spf check hard fail
 - [ ] spf fails with out email address
- [ ] spf check pass

-----
- [ ] gui
 - [ ] gui resizes correctly
 - [ ] objects on the gui flow with the rest of the interface

**Comment by guitarmanj** [comment]
For hard fail - securestate.com
For soft fail - google/Gmail
No spf: any of our clients hahaha



Sent from my T-Mobile 4G LTE Device


-------- Original message --------

**Pull request merged**